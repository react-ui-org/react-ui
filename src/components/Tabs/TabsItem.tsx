import React from 'react';
import { withGlobalProps } from '../../providers/globalProps';
import { classNames } from '../../helpers/classNames';
import { transferProps } from '../../helpers/transferProps';
import styles from './TabsItem.module.scss';
import type { TabsItemProps } from './TabsItem.types';

export const TabsItem: React.FunctionComponent<TabsItemProps> = ({
  afterLabel,
  beforeLabel,
  href,
  id,
  isActive = false,
  label,
  onClick,
  ...restProps
}) => (
  <li
    className={classNames(
      styles.root,
      isActive && styles.isRootActive,
    )}
    id={id}
    key={href}
  >
    <a
      {...transferProps(restProps)}
      className={styles.link}
      href={href}
      id={id && `${id}__link`}
      onClick={onClick}
    >
      <span className={styles.linkContent}>
        {beforeLabel}
        <span
          className={styles.label}
          id={id && `${id}__label`}
        >
          {label}
        </span>
        {afterLabel}
      </span>
    </a>
  </li>
);

export const TabsItemWithGlobalProps = withGlobalProps(TabsItem, 'TabsItem');

export default TabsItemWithGlobalProps;
