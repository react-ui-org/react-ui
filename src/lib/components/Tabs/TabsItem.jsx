import PropTypes from 'prop-types';
import React from 'react';
import { withProviderContext } from '../../provider';
import { classNames } from '../../utils/classNames';
import styles from './TabsItem.scss';

export const TabsItem = ({
  afterLabel,
  beforeLabel,
  href,
  id,
  isActive,
  label,
  onClick,
}) => (
  <li
    className={classNames(
      styles.root,
      isActive && styles.isRootActive,
    )}
    id={id}
    key={href}
  >
    <a
      href={href}
      onClick={onClick}
      className={styles.link}
      id={id && `${id}__link`}
    >
      <span className={styles.linkContent}>
        {beforeLabel}
        <span
          className={styles.label}
          id={id && `${id}__label`}
        >
          {label}
        </span>
        {afterLabel}
      </span>
    </a>
  </li>
);

TabsItem.defaultProps = {
  afterLabel: null,
  beforeLabel: null,
  id: undefined,
  isActive: false,
  onClick: null,
};

TabsItem.propTypes = {
  /**
   * Optional element shown after item's label, e.g. an icon.
   */
  afterLabel: PropTypes.node,
  /**
   * Optional element shown before item's label, e.g. an icon.
   */
  beforeLabel: PropTypes.node,
  /**
   * Item's link URL. Optionally add custom routing function to the `onClick` option to bypass
   * browser's default navigation.
   */
  href: PropTypes.string.isRequired,
  /**
   * ID of the root HTML element. It also serves as base for nested elements:
   * * `<ID>__link`
   * * `<ID>__label`
   */
  id: PropTypes.string,
  /**
   * If `true`, item is marked as active.
   */
  isActive: PropTypes.bool,
  /**
   * Item's label.
   */
  label: PropTypes.string.isRequired,
  /**
   * Function to call on item click, e.g. custom routing function.
   */
  onClick: PropTypes.func,
};

export const TabsItemWithContext = withProviderContext(TabsItem, 'TabsItem');

export default TabsItemWithContext;
