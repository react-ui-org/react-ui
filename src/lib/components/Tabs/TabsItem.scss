// 1. Keep inactive items under Tabs' decorative bottom border and pop active ones above it.

@use "../../styles/tools/breakpoint";
@use "../../styles/tools/reset";
@use "../../styles/tools/transition";
@use "theme";

.root {
    @include transition.add(transform);

    position: relative; // 1.
    z-index: 1; // 1.
    flex: none;
    padding-top: calc(-1 * #{theme.$item-active-shift-y});
    margin-bottom: theme.$item-active-shift-y;
    transform: translateY(0);

    &:hover,
    &:focus-within {
        transform: translateY(theme.$item-hover-shift-y);
    }

    &:not(:last-child) {
        margin-right: theme.$gap-xs;
    }

    @include breakpoint.up(sm) {
        &:not(:last-child) {
            margin-right: theme.$gap-sm;
        }
    }

    @include breakpoint.up(md) {
        &:not(:last-child) {
            margin-right: theme.$gap-md;
        }
    }
}

.link {
    @include reset.link();
    @include transition.add((color, border-color, background-color, box-shadow));

    display: block;
    padding: theme.$item-padding-xs;
    font-weight: theme.$item-font-weight;
    line-height: 1;
    text-decoration: none;
    color: theme.$item-color;
    border-width: theme.$item-border-width;
    border-style: solid;
    border-color: theme.$item-border-color;
    border-top-left-radius: theme.$item-border-radius;
    border-top-right-radius: theme.$item-border-radius;
    background-color: theme.$item-background-color;
    box-shadow: theme.$item-box-shadow;

    &:hover,
    &:focus {
        font-weight: theme.$item-hover-font-weight;
        color: theme.$item-hover-color;
        border-width: theme.$item-hover-border-width;
        border-color: theme.$item-hover-border-color;
        background-color: theme.$item-hover-background-color;
        box-shadow: theme.$item-hover-box-shadow;
    }

    @include breakpoint.up(sm) {
        padding: theme.$item-padding-sm;
    }

    @include breakpoint.up(md) {
        padding: theme.$item-padding-md;
    }
}

.linkContent {
    @include transition.add(transform);

    display: flex;
    align-items: center;
    transform: translateY(0);
}

.label:not(:first-child) {
    margin-left: theme.$item-icon-gap;
}

.label:not(:last-child) {
    margin-right: theme.$item-icon-gap;
}

.isRootActive,
.isRootActive:hover,
.isRootActive:focus-within {
    z-index: 2; // 1.
    transform: translateY(theme.$item-active-shift-y);
}

.isRootActive .link {
    font-weight: theme.$item-active-font-weight;
    color: theme.$item-active-color;
    border-width: theme.$item-active-border-width;
    border-color: theme.$item-active-border-color;
    background-color: theme.$item-active-background-color;
    box-shadow: theme.$item-active-box-shadow;
}

.link:hover .linkContent,
.link:focus .linkContent {
    transform: translateY(theme.$item-hover-label-shift-y);
}

.isRootActive .link .linkContent,
.isRootActive .link:hover .linkContent,
.isRootActive .link:focus .linkContent {
    transform: translateY(theme.$item-active-label-shift-y);
}
