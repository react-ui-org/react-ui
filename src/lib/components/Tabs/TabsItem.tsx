import React from 'react';
import { withGlobalProps } from '../../provider';
import { transferProps } from '../_helpers/transferProps';
import { classNames } from '../../utils/classNames';
import styles from './TabsItem.scss';
import { TabsItemProps } from './Tabs.types';

export const TabsItem: React.FunctionComponent<TabsItemProps> = ({
  afterLabel,
  beforeLabel,
  href,
  id,
  isActive = false,
  label,
  onClick,
  ...restProps
}) => (
  <li
    {...transferProps(restProps)}
    className={classNames(
      styles.root,
      isActive && styles.isRootActive,
    )}
    id={id}
    key={href}
  >
    <a
      className={styles.link}
      href={href}
      id={id && `${id}__link`}
      onClick={onClick}
    >
      <span className={styles.linkContent}>
        {beforeLabel}
        <span
          className={styles.label}
          id={id && `${id}__label`}
        >
          {label}
        </span>
        {afterLabel}
      </span>
    </a>
  </li>
);

export const TabsItemWithGlobalProps = withGlobalProps(TabsItem, 'TabsItem');

export default TabsItemWithGlobalProps;
