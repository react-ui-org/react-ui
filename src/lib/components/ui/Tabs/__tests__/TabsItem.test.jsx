import React from 'react';
import {
  mount,
  shallow,
} from 'enzyme';
import sinon from 'sinon';
import { TabsItem } from '../TabsItem';

describe('rendering', () => {
  it('renders correctly with mandatory props', () => {
    const tree = shallow((
      <TabsItem
        href="/tab-1"
        label="Tab 1"
      />
    ));

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with all props', () => {
    const tree = shallow((
      <TabsItem
        afterLabel={(<span>after label</span>)}
        beforeLabel={(<span>before label</span>)}
        href="/tab-1"
        id="my-tab"
        isActive
        label="Tab 1"
        onClick={() => {}}
      />
    ));

    expect(tree).toMatchSnapshot();
  });
});

describe('functionality', () => {
  it('calls onClick() on navigating', () => {
    const spy = sinon.spy();
    const component = mount((
      <TabsItem
        href="/tab-1"
        label="Tab 1"
        onClick={spy}
      />
    ));

    component.find('a').at(0).simulate('click');
    expect(spy.calledOnce).toEqual(true);
  });
});
