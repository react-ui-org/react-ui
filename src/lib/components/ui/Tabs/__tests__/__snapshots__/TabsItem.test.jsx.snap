// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly with all props 1`] = `
<li
  className="root isRootActive"
  id="my-tab"
>
  <a
    className="link"
    href="/tab-1"
    id="my-tab__link"
    onClick={[Function]}
  >
    <span
      className="linkContent"
    >
      <span>
        before label
      </span>
      <span
        className="label"
        id="my-tab__label"
      >
        Tab 1
      </span>
      <span>
        after label
      </span>
    </span>
  </a>
</li>
`;

exports[`rendering renders correctly with mandatory props 1`] = `
<li
  className="root "
>
  <a
    className="link"
    href="/tab-1"
    onClick={null}
  >
    <span
      className="linkContent"
    >
      <span
        className="label"
      >
        Tab 1
      </span>
    </span>
  </a>
</li>
`;
